File: apps/api-rs/src/api/runtime_events.rs
use axum::http::HeaderMap;
use axum::response::sse::Event;
use serde::Serialize;
use serde_json::{json, Value};
use uuid::Uuid;

use crate::models::now;

pub const EVENT_PROTOCOL_HEADER: &str = "x-documind-event-protocol";
pub const ATOM_SCHEMA_VERSION: &str = "moss.execution.event.v1";

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SseProtocol {
    Legacy,
    Atom,
}

impl SseProtocol {
    pub fn from_headers(headers: &HeaderMap) -> Self {
        let requested = headers
            .get(EVENT_PROTOCOL_HEADER)
            .and_then(|value| value.to_str().ok())
            .map(str::trim);

        match requested {
            Some("atom") | Some(ATOM_SCHEMA_VERSION) => SseProtocol::Atom,
            _ => SseProtocol::Legacy,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct RuntimeStep {
    pub step_id: String,
    pub parent_step_id: Option<String>,
    pub step_type: String,
    pub name: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct RuntimeEventEnvelope {
    pub schema_version: &'static str,
    pub event_id: String,
    pub job_id: Uuid,
    pub tenant_id: Uuid,
    pub user_id: Uuid,
    pub agent_id: String,
    pub session_id: Uuid,
    pub execution_id: Uuid,
    pub event_seq: u64,
    pub event_type: String,
    pub occurred_at: chrono::DateTime<chrono::Utc>,
    pub response_message_id: Uuid,
    pub trace_id: String,
    pub step: Option<RuntimeStep>,
    pub payload: Value,
}

#[derive(Debug, Clone)]
pub struct RuntimeEventFactory {
    job_id: Uuid,
    tenant_id: Uuid,
    user_id: Uuid,
    session_id: Uuid,
    execution_id: Uuid,
    response_message_id: Uuid,
    trace_id: String,
    next_seq: u64,
}

impl RuntimeEventFactory {
    pub fn new(
        tenant_id: Uuid,
        user_id: Uuid,
        session_id: Uuid,
        response_message_id: Uuid,
    ) -> Self {
        let job_id = Uuid::new_v4();
        Self {
            job_id,
            tenant_id,
            user_id,
            session_id,
            execution_id: Uuid::new_v4(),
            response_message_id,
            trace_id: format!("trace_{job_id}"),
            next_seq: 1,
        }
    }

    pub fn event(&mut self, event_type: &str, payload: Value) -> Event {
        self.event_with_step(event_type, None, payload)
    }

    pub fn event_with_step(
        &mut self,
        event_type: &str,
        step: Option<RuntimeStep>,
        payload: Value,
    ) -> Event {
        let seq = self.next_seq;
        self.next_seq += 1;
        let envelope = RuntimeEventEnvelope {
            schema_version: ATOM_SCHEMA_VERSION,
            event_id: format!("evt_{}_{}", self.job_id.simple(), seq),
            job_id: self.job_id,
            tenant_id: self.tenant_id,
            user_id: self.user_id,
            agent_id: "documind_default".to_string(),
            session_id: self.session_id,
            execution_id: self.execution_id,
            event_seq: seq,
            event_type: event_type.to_string(),
            occurred_at: now(),
            response_message_id: self.response_message_id,
            trace_id: self.trace_id.clone(),
            step,
            payload,
        };

        Event::default()
            .event(event_type)
            .id(envelope.event_id.clone())
            .data(json!(envelope).to_string())
    }
}

pub fn tool_step(tool_call_id: &str, name: &str) -> RuntimeStep {
    RuntimeStep {
        step_id: tool_call_id.to_string(),
        parent_step_id: None,
        step_type: "ToolCall".to_string(),
        name: name.to_string(),
    }
}

File: apps/api-bun/src/api/runtime_events.ts
// 移植自 apps/api-rs/src/api/runtime_events.rs —— Atom SSE 事件协议
import { newUuid } from '../infra/uuid.ts';
import { nowRfc3339 } from '../infra/time.ts';

export const EVENT_PROTOCOL_HEADER = 'x-documind-event-protocol';
export const ATOM_SCHEMA_VERSION = 'moss.execution.event.v1';

export type SseProtocol = 'legacy' | 'atom';

export function sseProtocolFromHeaders(headers: Headers): SseProtocol {
  const requested = headers.get(EVENT_PROTOCOL_HEADER)?.trim();
  return requested === 'atom' || requested === ATOM_SCHEMA_VERSION ? 'atom' : 'legacy';
}

export interface RuntimeStep {
  step_id: string; parent_step_id: string | null;
  step_type: string; name: string;
}

export interface RuntimeEventEnvelope {
  schema_version: string; event_id: string; job_id: string; tenant_id: string;
  user_id: string; agent_id: string; session_id: string; execution_id: string;
  event_seq: number; event_type: string; occurred_at: string;
  response_message_id: string; trace_id: string;
  step: RuntimeStep | null; payload: unknown;
}

export interface SseEvent { event: string; id: string; data: string; }

export class RuntimeEventFactory {
  private readonly jobId: string;
  private readonly tenantId: string;
  private readonly userId: string;
  private readonly sessionId: string;
  private readonly executionId: string;
  private readonly responseMessageId: string;
  private readonly traceId: string;
  private nextSeq = 1;

  constructor(tenantId: string, userId: string, sessionId: string, responseMessageId: string) {
    this.jobId = newUuid();
    this.tenantId = tenantId;
    this.userId = userId;
    this.sessionId = sessionId;
    this.executionId = newUuid();
    this.responseMessageId = responseMessageId;
    this.traceId = `trace_${this.jobId}`;
  }

  event(eventType: string, payload: unknown): SseEvent {
    return this.eventWithStep(eventType, null, payload);
  }

  eventWithStep(eventType: string, step: RuntimeStep | null, payload: unknown): SseEvent {
    const seq = this.nextSeq;
    this.nextSeq += 1;
    const envelope: RuntimeEventEnvelope = {
      schema_version: ATOM_SCHEMA_VERSION,
      event_id: `evt_${this.jobId.replace(/-/g, '')}_${seq}`,
      job_id: this.jobId, tenant_id: this.tenantId, user_id: this.userId,
      agent_id: 'documind_default', session_id: this.sessionId,
      execution_id: this.executionId, event_seq: seq, event_type: eventType,
      occurred_at: nowRfc3339(), response_message_id: this.responseMessageId,
      trace_id: this.traceId, step, payload,
    };
    return { event: eventType, id: envelope.event_id, data: JSON.stringify(envelope) };
  }
}

export function toolStep(toolCallId: string, name: string): RuntimeStep {
  return {
    step_id: toolCallId, parent_step_id: null, step_type: 'ToolCall', name,
  };
}
